azure-iot-device/tests/common/pipeline/pipeline_data_object_test.py

```diff
@@ -0,0 +1,134 @@
+):
+    """
+    Add a test class to test the given PipelineOperation class.  The class that
+    we're testing is passed in the obj parameter, and the different initialization
+    constants are passed with the named arguments that follow.
+    """
+    all_extra_defaults = extra_defaults.copy()
+    all_extra_defaults.update(name=obj.__name__)
+
+    _add_data_object_test(
+        obj=obj,
+        module=module,
+        base_defaults=base_operation_defaults,
+        extra_defaults=all_extra_defaults,
+        positional_arguments=positional_arguments,
+        keyword_arguments=keyword_arguments,
+    )
+
+
+def add_event_test(obj, module, extra_defaults={}, positional_arguments=[], keyword_arguments={}):
+    """
+    Add a test class to test the given PipelineOperation class.  The class that
+    we're testing is passed in the obj parameter, and the different initialization
+    constants are passed with the named arguments that follow.
+    """
+    all_extra_defaults = extra_defaults.copy()
+    all_extra_defaults.update(name=obj.__name__)
+
+    _add_data_object_test(
+        obj=obj,
+        module=module,
+        base_defaults=base_event_defaults,
+        extra_defaults=all_extra_defaults,
+        positional_arguments=positional_arguments,
+        keyword_arguments=keyword_arguments,
+    )
+
+
+def _add_data_object_test(
+    obj, module, base_defaults, extra_defaults={}, positional_arguments=[], keyword_arguments={}
+):
+    """
+    internal function that takes the class and attribute details and adds a test class which
+    validates that the given class properly implements the given attributes.
+    """
+
+    # `defaults` contains an array of object attributes that should be set when
+    # we call the initializer will all of the required positional arguments
+    # and none of the optional keyword arguments.
+
+    defaults = base_defaults.copy()
+    for key in extra_defaults:
+        defaults[key] = extra_defaults[key]
+    for key in keyword_arguments:
+        defaults[key] = keyword_arguments[key]
+
+    # `args` contains an array of positional argument that we are passing to test that they
+    # get assigned to the correct attribute.
+    args = [get_next_fake_value() for i in range(len(positional_arguments))]
+
+    # `kwargs` contains a dictionary of all keyword arguments, which includes required positional
+    # arguments and optional keyword arguments.
+    kwargs = {}
+    for key in positional_arguments:
+        kwargs[key] = get_next_fake_value()
+    for key in keyword_arguments:
+        kwargs[key] = get_next_fake_value()
+
+    # LocalTestObject is a local class which tests the object that was passed in.  pytest doesn't test
+    # against this local object, but it does test against it when we put it into the module namespace
+    # for the module that was passed in.
+    @pytest.mark.describe("{} - Instantiation".format(obj.__name__))
```

Comment:
I would argue that since these are the only tests for the Event/Ops, that you don't even need to specify "Instantiation" here.

Reply:
Fair enough.  Does that change the phrasing of the @it clauses?  Instantiation accepts positional arguments, but is it correct to say the object accepts positional arguments?  I'm having trouble with the phrasing here.  

---
In reply to: [291679669](https://github.com/Azure/azure-iot-sdk-python-preview/pull/104#discussion_r291679669) [](ancestors = 291679669)

Reply:
Actually, never mind.  I'm leaving this because I have some code-reuse in mind that works better if we leave this.

---
In reply to: [291699321](https://github.com/Azure/azure-iot-sdk-python-preview/pull/104#discussion_r291699321) [](ancestors = 291699321,291679669)

Reply:
Okay cool. 

---
In reply to: [291705442](https://github.com/Azure/azure-iot-sdk-python-preview/pull/104#discussion_r291705442) [](ancestors = 291705442,291699321,291679669)